src/components/Messages/MessageIcon.tsx

```diff
@@ -0,0 +1,93 @@
+import useXmtpClient from '@components/utils/hooks/useXmtpClient';
+import { MailIcon } from '@heroicons/react/outline';
+import conversationMatchesProfile from '@lib/conversationMatchesProfile';
+import { fetchMostRecentMessage } from '@lib/fetchMostRecentMessage';
+import type { DecodedMessage } from '@xmtp/xmtp-js';
+import Link from 'next/link';
+import type { FC } from 'react';
+import { useEffect } from 'react';
+import { useAppPersistStore, useAppStore } from 'src/store/app';
+
+const MessageIcon: FC = () => {
+  const currentProfile = useAppStore((state) => state.currentProfile);
+  const { client: cachedClient } = useXmtpClient(true);
+  const clearMessagesBadge = useAppPersistStore((state) => state.clearMessagesBadge);
+  const lastViewedMessagesAt = useAppPersistStore((state) =>
+    state.lastViewedMessagesAt ? new Date(state.lastViewedMessagesAt) : null
+  );
+  const showUnreadMessages = useAppPersistStore((state) => state.showUnreadMessages);
+  const setShowUnreadMessages = useAppPersistStore((state) => state.setShowUnreadMessages);
+
+  const showBadge = (mostRecentMessage: DecodedMessage | undefined): boolean => {
+    if (!mostRecentMessage) {
+      return false;
+    }
+    return !lastViewedMessagesAt || lastViewedMessagesAt.getTime() < mostRecentMessage.sent.getTime();
+  };
+
+  useEffect(() => {
+    if (!cachedClient || !currentProfile) {
+      return;
+    }
+
+    const matcherRegex = conversationMatchesProfile(currentProfile.id);
+
+    const fetchShowBadge = async () => {
+      const convos = await cachedClient.conversations.list();
+      const matchingConvos = convos.filter(
+        (convo) => convo.context?.conversationId && matcherRegex.test(convo.context.conversationId)
+      );
+      if (matchingConvos.length <= 0) {
+        return;
+      }
+
+      const previews = await Promise.all(matchingConvos.map(fetchMostRecentMessage));
```

Comment:
The one thing that's a little scary here is the number of API requests involved on every pageview. If you have 100 people who have ever messaged you, this will lead to 101 API requests just to show the badge. 

To just get the timestamp of the most recent message, the more performant option would be to do something like.

```ts
const topics = matchingConvos.map((convo) => convo.topic)
const recentMessage = client.listEnvelopes([], (env) => env.timestampNs, { limit: 1, direction: SortDirection.SORT_DIRECTION_DESCENDING})
const lastMessageTimestamp = recentMessages.length ? somehowConvertNanosecondTimestampToDate(recentMessages[0]) : null
```

Reply:
Nick!!! This is brilliant. Yeah, I wrote down today that we'll likely soon want to be able to paginate & sort by updatedAt for convos 😅 (no idea at how feasible that would be today), but this is so much better for this use case. Updating!